### PR TITLE
 development/pe 51 spec module

### PR DESCRIPTION
**Pull request by davidlindley**

### CODE REVIEW

File: spec-modules/oj-order-summary.spec-module.js [hunk 1]
```diff
@@ -33,9 +36,9 @@ module.exports = function(config) {
       var elemsSectionHeaders = element.all(by.css('.sectionTitle.priceTableShort')),
         elemsSectionBodies = element.all(by.css('.contentBlock.priceTableDetail'));
 
-      elemsSectionHeaders.each(function(elemHeader, index) {
+      elemsSectionHeaders.each(function(elemHeader, sectionIndex) {
```
Comment (danciu) [inline spec-modules/oj-order-summary.spec-module.js]: `index` to `sectionIndex` makes it more readable and descriptive 👍

File: spec-modules/oj-order-summary.spec-module.js [hunk 2]
```diff
@@ -47,19 +50,47 @@ module.exports = function(config) {
         // h2.aside
         expect(elemHeading.getText()).toBe(configSection.name);
 
-        // Get all section bodies
-        var elemsSectionBodyHeadings = elemsSectionBodies.get(index).$$('h4');
+        // Get all section body headings and descriptions
+        var elemsSectionBodyHeadings = elemsSectionBodies.get(sectionIndex).$$('h4');
+        var elemsSectionBodyDescription = elemsSectionBodies.get(sectionIndex).$$('.description');
+        var elementSectionListContainer = elemsSectionBodies.get(sectionIndex).$$('ul.bulletList');
```
Comment (danciu) [inline spec-modules/oj-order-summary.spec-module.js]: `elemsSectionBodyDescription` and `elementSectionListContainer` are used only in their optional sections, so the caching is a bit unnecessary in this case. We could instead cache `elemsSectionBodies.get(sectionIndex)` here and then use them in their sections accordingly:

```javascript
elemSectionIndex.$$('h4').each(function(sectionBodyHeading, subSectionIndex) {
```

```javascript
expect(elemSectionIndex.$$('.description').get(subSectionIndex).getText()).toEqual(itemToTest.description);
```

I do agree that's a bit harder to read, so it's up to you, I am fine with both. 👍 

If we leave them like that, can you please separate the variable declarations by comma and group them together please? This is in order to maintain consistency within the file, as we are grouping common variable declarations.

Thanks

File: spec-modules/oj-order-summary.spec-module.js [hunk 3]
```diff
@@ -47,19 +50,47 @@ module.exports = function(config) {
...
+
+            elementSectionListContainer.each(function(listContainer, outerListIndex) {
+              var listItems = listContainer.$$('li');
+              listItems.each(function(listItem, innerListIndex) {
+
+                listItem.getText().then(function(listText) {
+                  if(listText.length > 0) {
+                    expect(listText).toBe(itemToTest.list[listLoopCount]);
+                  }
+                  listLoopCount++;
```
Comment (danciu) [inline spec-modules/oj-order-summary.spec-module.js]: Doesn't `listLoopCount` always have the same value as `innerListIndex` in this case?
First iteration is `0` and `0` and both counters get incremented each iteration, no matter the text length (the conditional above the increment).

File: spec-modules/oj-order-summary.spec-module.js [hunk 4]
```diff
@@ -5,7 +5,10 @@
  *     @param {string} section.name Name of section in array
  *     @param {string} section.oneoff One-off value for the section
  *     @param {string} section.monthly Monthly value for the section
- *     @param {array} section.items Array with text content of section items
+ *     @param {array} section.items Array of item Objects
+          @param {string} item.title Title of the subSection
+          @param {string} item.description Description of the subSection
+          @param {array} item.list array of string elements
```
Comment (danciu) [inline spec-modules/oj-order-summary.spec-module.js]: 👍

File: spec-modules/oj-single-option-details.spec-module.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
+/**
+ * OJ Single Option Details
+ * @param  {Object} config Config object
+ *   @param {array} config.sections Array with section Objects
+        @param {string} section.heading. Heading of the section
+        @param {array} config.sections.subsection Array with subsection object
+     *     @param {string} subsection.name Name of option name
+     *     @param {string} subsection.description Desription of the option
+     *     @param {string} subsection.oneOff One off cost for the item (optional)
+     *     @param {array} subsection.Monthly Array with text content of section items (optional)
```
Comment (danciu) [inline spec-modules/oj-single-option-details.spec-module.js]: Lowe case `m` 😛

File: spec-modules/oj-single-option-details.spec-module.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  return it('Finished checking the OJ Single Option Details', function() {
+    var sections = config.sections;
+
+    // In-module information, display start heading
+    console.log(format.formatInModuleStartInfo('Started checking the Single Option Details...'));
+
+    if (sections && sections.length) {
+      var sectionContainers = element.all(by.css(''))
```
Comment (danciu) [inline spec-modules/oj-single-option-details.spec-module.js]: ❌ Pending delete:D

File: spec-modules/oj-single-option-details.spec-module.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+      elemsSections.each(function(section, sectionIndex) {
+        var heading = section.$('.headingH3'),
+          tableBody = elemsSections.get(sectionIndex).$('tbody'),
+          tableRows = tableBody.$$('tr'),
+          section = sections[sectionIndex],
+          sectionHeader = section.header,
+          sectionSubSections = section.subsections;
+
+          if(section) {
```
Comment (danciu) [inline spec-modules/oj-single-option-details.spec-module.js]: We can optimise readability for this `if..else` loop using `if` and early `return`:

```javascript
if (!subSection) {
  console.log(format.formatError('No test data for all the subsections within ' +  sectionHeader));
  
  // Early return
  return;
}

// subSection exists code here
```

File: spec-modules/oj-single-option-details.spec-module.js [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+            expect(heading.getText()).toBe(sectionHeader);
+
+            tableRows.each(function(row, rowIndex) {
+              var subSection = section.subsections[rowIndex],
+                name = tableRows.get(rowIndex).$('.col_info label'),
+                description = tableRows.get(rowIndex).$('.col_info p'),
+                oneOffCost = tableRows.get(rowIndex).$('.col_oneOffCost'),
+                monthlyCost = tableRows.get(rowIndex).$('.col_monthlyCost');
+
+                if(subSection) {
```
Comment (danciu) [inline spec-modules/oj-single-option-details.spec-module.js]: We can optimise readability for this `if..else` loop using `if` and early `return`:

```javascript
if (!subSection) {
  console.log(format.formatError('No test data for all the subsections within ' +  sectionHeader));
  
  // Early return
  return;
}

// subSection exists code here
```

File: spec-modules/oj-single-option-details.spec-module.js [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+          if(section) {
+            // Check the section header
+            expect(heading.getText()).toBe(sectionHeader);
+
+            tableRows.each(function(row, rowIndex) {
+              var subSection = section.subsections[rowIndex],
+                name = tableRows.get(rowIndex).$('.col_info label'),
+                description = tableRows.get(rowIndex).$('.col_info p'),
+                oneOffCost = tableRows.get(rowIndex).$('.col_oneOffCost'),
+                monthlyCost = tableRows.get(rowIndex).$('.col_monthlyCost');
```
Comment (danciu) [inline spec-modules/oj-single-option-details.spec-module.js]: Some of the names for these elements suggest they may be holding a string value, not an object pointing to an element. Something more suggestive would be the same format you used above, adding an `elem` or `elems` prefix, e.g. `elemName`, `elemDescription` etc.

File: spec-modules/oj-single-option-details.spec-module.js [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+                  }
+
+                  // Check the one off costs
+                  if(subSection.oneOffCost) {
+                    expect(oneOffCost.getText()).toBe(subSection.oneOffCost);
+                  }
+
+                  // Check the monthly costs
+                  if(subSection.monthlyCost) {
+                    expect(monthlyCost.getText()).toBe(subSection.monthlyCost);
```
Comment (danciu) [inline spec-modules/oj-single-option-details.spec-module.js]: I really like how you used `expect` here, the section is really readable and clean.. and I know some of these files tend to get weird...so nice job! 💯

File: spec-modules/oj-single-option-details.spec-module.js [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+                  }
+                } else {
+                  console.log(format.formatError('No test data for all the subsections within ' +  sectionHeader));
+                }
+            });
+          } else {
+            console.log(format.formatError('No test data for section: '  + sectionHeader));
+          }
```
Comment (danciu) [inline spec-modules/oj-single-option-details.spec-module.js]: Same `if`/`else` recommendations from above if you fancy and think it does make code more readable. 👍

**Comment by danciu** [comment]
Really cool, @davidlindley. Great job, you got used to the framework and libraries really fast:D
I am so pleased to see the project get some love! ❤️

**Comment by danciu** [comment]
🍻

File: spec-modules/oj-order-summary.spec-module.js (hunk 2 above)
Comment (davidlindley) [inline spec-modules/oj-order-summary.spec-module.js]: I think for readability I will do the second option.

File: spec-modules/oj-order-summary.spec-module.js (hunk 3 above)
Comment (davidlindley) [inline spec-modules/oj-order-summary.spec-module.js]: I don't think so. The listLoopCount is incremented for every inner loop itteration, but maintains its value as the outloop increments as well. So 

Outter Loop 0 : Inner loop 0,1,2,3 (listLoopCount == 4)
Outer Loop 1: Inner loop 0,1,2,3 (listLoopCount == 8)

Or something like that. This is to allow for simplifying the JSON object.

Comment (danciu) [inline spec-modules/oj-order-summary.spec-module.js]: Ah, you want it incremented for both. Ok, coolio 👍

**Comment by danciu** [comment]
💯

**Pull request merged**